lib/spl/src/spl_emit.c

```diff
@@ -0,0 +1,347 @@
+#include "nitro/types.h"
+#include <nitro/fx/fx.h>
+#include <nitro/fx/fx_const.h>
+#include <nitro/fx/fx_trig.h>
+
+#include "spl_emitter.h"
+#include "spl_internal.h"
+#include "spl_list.h"
+#include "spl_particle.h"
+#include "spl_random.h"
+#include "spl_resource.h"
+
+static void SPLEmitter_ComputeOrthogonalAxes(SPLEmitter *emtr);
+static void SPLUtil_TiltCoordinates(VecFx32 *ptclPos, VecFx32 *pos, SPLEmitter *emtr);
+
+static VecFx16 sUpVector = { 0, FX16_ONE, 0 };
+
+static void SPLEmitter_ComputeOrthogonalAxes(SPLEmitter *emtr)
+{
+    VecFx16 vec, axis;
+
+    vec = sUpVector;
+
+    switch (emtr->resource->header->flags.circleAxis) {
+    case SPL_CIRCLE_AXIS_X:
+        axis.x = FX32_ONE;
+        axis.y = 0;
+        axis.z = 0;
+        break;
+    case SPL_CIRCLE_AXIS_Y:
+        axis.x = 0;
+        axis.y = FX32_ONE;
+        axis.z = 0;
+        break;
+    case SPL_CIRCLE_AXIS_Z:
+        axis.x = 0;
+        axis.y = 0;
+        axis.z = FX32_ONE;
+        break;
+    default: // SPL_CIRCLE_AXIS_EMITTER
+        VEC_Fx16Normalize(&emtr->axis, &axis);
+        break;
+    }
+
+    fx32 dot = VEC_Fx16DotProduct(&vec, &axis);
+    if (dot == FX32_ONE || dot == -FX32_ONE) {
+        vec.x = FX32_ONE;
+        vec.y = 0;
+        vec.z = 0;
+    }
+
+    emtr->crossAxis1.x = FX_MUL(axis.y, vec.z) - FX_MUL(axis.z, vec.y);
+    emtr->crossAxis1.y = FX_MUL(axis.z, vec.x) - FX_MUL(axis.x, vec.z);
+    emtr->crossAxis1.z = FX_MUL(axis.x, vec.y) - FX_MUL(axis.y, vec.x);
+
+    emtr->crossAxis2.x = FX_MUL(axis.y, emtr->crossAxis1.z) - FX_MUL(axis.z, emtr->crossAxis1.y);
+    emtr->crossAxis2.y = FX_MUL(axis.z, emtr->crossAxis1.x) - FX_MUL(axis.x, emtr->crossAxis1.z);
+    emtr->crossAxis2.z = FX_MUL(axis.x, emtr->crossAxis1.y) - FX_MUL(axis.y, emtr->crossAxis1.x);
+
+    VEC_Fx16Normalize(&emtr->crossAxis1, &emtr->crossAxis1);
+    VEC_Fx16Normalize(&emtr->crossAxis2, &emtr->crossAxis2);
+}
+
+// Tilt the coordinates of a particle based on the emitter's axis
+static void SPLUtil_TiltCoordinates(VecFx32 *ptclPos, VecFx32 *pos, SPLEmitter *emtr)
+{
+    VecFx16 vec;
+    VEC_Fx16CrossProduct(&emtr->crossAxis1, &emtr->crossAxis2, &vec);
+    VEC_Fx16Normalize(&vec, &vec);
+
+    ptclPos->x = FX_MUL(pos->x, emtr->crossAxis1.x) + FX_MUL(pos->y, emtr->crossAxis2.x) + FX_MUL(pos->z, vec.x);
+    ptclPos->y = FX_MUL(pos->x, emtr->crossAxis1.y) + FX_MUL(pos->y, emtr->crossAxis2.y) + FX_MUL(pos->z, vec.y);
+    ptclPos->z = FX_MUL(pos->x, emtr->crossAxis1.z) + FX_MUL(pos->y, emtr->crossAxis2.z) + FX_MUL(pos->z, vec.z);
+}
+
+void SPLEmitter_EmitParticles(SPLEmitter *emtr, SPLParticleList *list)
+{
+    SPLResource *res = emtr->resource;
+    SPLResourceHeader *header = res->header;
+
+    int i = 0;
```

Comment:
**question:** Is this declaration required here? It doesn't seem to be used before the re-assignment in the header of the below for-loop.

Reply:
Unfortunately it doesn't match if I declare it in the for loop